server/market/market.go

```diff
@@ -1175,6 +1185,41 @@ func (m *Market) epochStart(orders []order.Order) (cSum []byte, ordersRevealed [
 	return
 }
 
+// Unbook is a callback for the swapper to request unbooking an order.
+func (m *Market) Unbook(lo *order.LimitOrder) bool {
+	// Ensure we do not unbook during matching.
+	m.bookMtx.Lock()
+	_, removed := m.book.Remove(lo.ID())
+	m.bookMtx.Unlock()
+
+	m.unlockOrderCoins(lo)
+
+	if !removed {
+		return false
+	}
+
+	// Create the server-generated cancel order, and register it with
+	// the AuthManager for cancellation ratio computation.
+	coid, revTime, err := m.storage.RevokeOrder(lo)
+	if err == nil {
+		m.auth.RecordCancel(lo.User(), coid, lo.ID(), revTime)
+	} else {
+		log.Errorf("Failed to revoke order %v with a new cancel order: %v",
+			lo.UID(), err)
+	}
+
+	// Send "unbook" notification to order book subscribers.
+	m.sendToFeeds(&updateSignal{
+		action: unbookAction,
+		data: sigDataUnbookedOrder{
+			order:    lo,
+			epochIdx: -1, // NOTE: no epoch
+		},
+	})
+
+	return true
```

Comment:
Other than epoch processing and matching, this is the only way the book can be modified.  The alternative is no good -- pushing unbook request from the Swapper into the epoch processing pipeline via cancel orders is very indirect and also permits more matches with a canceled order.